File: frontend/lib/services/caseService.ts
import { searchOpinions, getClusterJudges } from '@/lib/courtlistener';
import { generateText } from '@/lib/gemini';
import { getDb } from '@/lib/mongo';
import crypto from 'crypto';

export function caseId(name: string): string
{
	return crypto.createHash('sha1').update(name.toLowerCase().trim()).digest('hex').slice(0, 10);
}

export async function expandSummary(name: string, snippet: string): Promise<string>
{
	const prompt = `Summarize the US Supreme Court case "${name}" in 3-4 paragraphs.
Cover: the facts and background, the central legal question, the Court's holding, and the broader significance.
Write in clear informative prose. No headers or bullet points.
Context from the opinion: ${snippet}`;
	try
	{
		return await generateText(prompt);
	}
	catch
	{
		return snippet;
	}
}

export async function mapCl(
	result: any,
	options?: { overrideId?: string; expand?: boolean }
): Promise<any | null>
{
	const name = (result.caseName ?? '').trim();
	if(!name) return null;

	const expand = options?.expand ?? true;

	let year: number | null = null;
	if(result.dateFiled)
	{
		const parsed = parseInt(result.dateFiled.slice(0, 4), 10);
		if(!isNaN(parsed)) year = parsed;
	}

	const citations: string[] = result.citation ?? [];
	const snippet = (result.snippet ?? '').replace(/<[^>]+>/g, '').trim();
	const rawUrl: string = result.absolute_url ?? '';
	const judgeStr: string = result.judge ?? '';

	let panel: string[] | null = null;
	if(!judgeStr && rawUrl)
	{
		panel = await getClusterJudges(rawUrl);
	}
	else
	{
		const names = judgeStr.split(',').map((n: string) => n.trim()).filter(Boolean);
		panel = names.length > 0 ? names : null;
	}

	const summary = expand ? await expandSummary(name, snippet) : snippet;

	return {
		id: options?.overrideId ?? caseId(name),
		name,
		year,
		category: result.suitNature || 'Supreme Court',
		summary,
		citation: citations[0] ?? '',
		court_listener_link: rawUrl ? `https://www.courtlistener.com${rawUrl}` : null,
		panel,
	};
}

export async function clSearch(query: string, limit = 5): Promise<any[]>
{
	const results = await searchOpinions(query, { limit });

	const unexpanded = await Promise.all(
		results.map((r: any) => mapCl(r, { expand: false }))
	);
	const valid = unexpanded.filter(Boolean) as any[];

	const expanded = await Promise.all(
		valid.map((c: any) => expandSummary(c.name, c.summary).then(summary => ({ ...c, summary })))
	);

	const seen = new Set<string>();
	const deduped = expanded.filter((c: any) =>
	{
		if(seen.has(c.id)) return false;
		seen.add(c.id);
		return true;
	});

	const db = await getDb();
	const col = db.collection('cases');
	await Promise.all(
		deduped.map((c: any) => col.replaceOne({ id: c.id }, c, { upsert: true }))
	);

	return deduped;
}

export async function seedIfEmpty(): Promise<void>
{
	const db = await getDb();
	const count = await db.collection('cases').countDocuments();
	if(count === 0)
	{
		const { run } = await import('@/lib/seed');
		await run();
	}
}

File: backend/app/services/case_service.py
from app.clients.courtlistener import CourtListenerClient
from concurrent.futures import ThreadPoolExecutor
from app.clients.gemini import GeminiClient
from app.clients.mongo import get_db
import hashlib
import re

EXPAND_WORKERS = 3


def _case_id(name: str) -> str:
    return hashlib.sha1(name.lower().strip().encode()).hexdigest()[:10]


def _expand_summary(name: str, snippet: str) -> str:
    prompt = f"""Summarize the US Supreme Court case "{name}" in 3-4 paragraphs.
Cover: the facts and background, the central legal question, the Court's holding, and the broader significance.
Write in clear informative prose. No headers or bullet points.
Context from the opinion: {snippet}"""
    try:
        return GeminiClient().generate(prompt)
    except Exception:
        return snippet


def _map_cl(result: dict, override_id: str | None = None, cl: CourtListenerClient | None = None, expand: bool = True) -> dict | None:
    name = result.get("caseName", "").strip()
    if not name:
        return None
    year = None
    if result.get("dateFiled"):
        try:
            year = int(result["dateFiled"][:4])
        except ValueError:
            pass
    citations = result.get("citation", [])
    snippet = re.sub(r"<[^>]+>", "", result.get("snippet", "")).strip()
    raw_url = result.get("absolute_url", "")

    judge_str = result.get("judge", "") or ""
    if not judge_str and raw_url and cl:
        panel = cl.get_cluster_judges(raw_url)
    else:
        names = [n.strip() for n in judge_str.split(",") if n.strip()]
        panel = names or None

    return {
        "id": override_id or _case_id(name),
        "name": name,
        "year": year,
        "category": result.get("suitNature") or "Supreme Court",
        "summary": _expand_summary(name, snippet) if expand else snippet,
        "citation": citations[0] if citations else "",
        "court_listener_link": f"https://www.courtlistener.com{raw_url}" if raw_url else None,
        "panel": panel,
    }


def _cl_search(query: str, limit: int = 5) -> list[dict]:
    cl = CourtListenerClient()
    try:
        results = cl.search_opinions(query, limit=limit)
        cases = [c for r in results if (c := _map_cl(r, cl=cl, expand=False))]

        def _expand(case: dict) -> dict:
            case["summary"] = _expand_summary(case["name"], case["summary"])
            return case

        with ThreadPoolExecutor(max_workers=EXPAND_WORKERS) as pool:
            cases = list(pool.map(_expand, cases))

        cases = list({c["id"]: c for c in cases}.values())

        col = get_db()["cases"]
        for c in cases:
            col.replace_one({"id": c["id"]}, c, upsert=True)
        return cases
    finally:
        cl.close()


def _seed_if_empty() -> None:
    if get_db()["cases"].count_documents({}) == 0:
        from app.seed import run
        run()
